pkg/lib/kitfile/generate/filesystem.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2025 The KitOps Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package generate
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+type DirectoryListing struct {
+	Name    string
+	Path    string
+	Files   []FileListing
+	Subdirs []DirectoryListing
+}
+
+type FileListing struct {
+	Name string
+	Path string
+	Size int64
+}
+
+func DirectoryListingFromFS(contextDir string) (*DirectoryListing, error) {
+	return genDirListingFromPath(".", contextDir)
+}
+
+func genDirListingFromPath(curDir, contextDir string) (*DirectoryListing, error) {
+	dirName := filepath.Base(curDir)
+	result := &DirectoryListing{
+		Name: dirName,
+		Path: curDir,
+	}
+
+	fullPath := filepath.Join(contextDir, curDir)
+	ds, err := os.ReadDir(fullPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read directory %s: %w", fullPath, err)
+	}
+	for _, dirEntry := range ds {
+		// Ignore anything that's not a directory or regular file
+		if !dirEntry.IsDir() && !dirEntry.Type().IsRegular() {
```

Comment:
You are checking `dirEntry.IsDir()` below and then assuming the else case is `IsRegular`. I think it would be cleaner to just have the check for type in one place. Otherwise things could get out of date and harder to maintain long term.

Reply:
Hm, I hadn't thought of it that way. In my mind, it's clearer to up-front exclude the types we're not interested in (links, devices, etc.), though I see your point in assuming `!isDir` is a regular file.

Reply:
Personally, I do this style when checking for things to ignore and then limit the nesting further down. For this, you do the check and then check again. If you want to better ignore the none supported types, I think you could do a switch statement over file types?

The way you have it works, so feel free to leave it if you find this way better.

Reply:
Fixed in 0759760